cmd/dep/init.go

```diff
@@ -210,6 +211,43 @@ func hasImportPathPrefix(s, prefix string) bool {
 	return strings.HasPrefix(s, prefix+"/")
 }
 
+// getProjectPropertiesFromVersion takes a gps.Version and returns a proper
+// gps.ProjectProperties with Constraint value based on the version type.
+func getProjectPropertiesFromVersion(v gps.Version) gps.ProjectProperties {
+	pp := gps.ProjectProperties{}
+	switch v.Type() {
+	case gps.IsBranch, gps.IsVersion, gps.IsRevision:
+		pp.Constraint = v
+	case gps.IsSemver:
+		c, _ := gps.NewSemverConstraint("^" + v.String())
+		pp.Constraint = c
+	}
+
+	return pp
+}
+
+// getSolverSolution runs gps solver and returns a solution.
+func getSolverSolution(root string, pkgT pkgtree.PackageTree, m *dep.Manifest, l *dep.Lock, sm *gps.SourceMgr) (gps.Solution, error) {
```

Comment:
I don't object to having this abstraction in principle, but let's leave it for a later PR where it can be a dedicated refactor that affects the other commands as well.